core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -628,9 +628,9 @@ class KafkaController(val config: KafkaConfig,
       topicDeletionManager.failReplicaDeletion(newOfflineReplicasForDeletion)
     }
 
-    // If replica failure did not require leader re-election, inform brokers of the offline brokers
+    // If no partitions are affected, inform brokers of the offline brokers
     // Note that during leader re-election, brokers update their metadata
-    if (partitionsWithOfflineLeader.isEmpty) {
+    if (newOfflineReplicas.isEmpty || (newOfflineReplicasNotForDeletion.isEmpty && partitionsWithOfflineLeader.isEmpty)) {
```

Comment:
I think that makes sense. Do we need to do something similar for both `partitionStateMachine` calls?
`handleStateChanges` already returns a `Map[TopicPartition, Either[Throwable, LeaderAndIsr]]` so it seems like we could just look to see if there are any values with `LeaderAndIsr` and that would imply that we sent out `updateMetadataRequest` as part of Online/Offline state changes. and likewise, we can have `triggerOnlinePartitionStateChange` return the same results.

I am thinking something along the lines of:
```
    // trigger OfflinePartition state for all partitions whose current leader is one amongst the newOfflineReplicas
    val offlineStateChangeResults = partitionStateMachine.handleStateChanges(partitionsWithOfflineLeader.toSeq, OfflinePartition)
    // trigger OnlinePartition state changes for offline or new partitions
    val onlineStateChangeResults = partitionStateMachine.triggerOnlinePartitionStateChange()

...
    val leaderElectionSucceeded = (offlineStateChangeResults ++ onlineStateChangeResults).find(_)
    // (!leaderElectionSucceeded && newOfflineReplicasForDeletion.isEmpty) {
      sendUpdateMetadataRequest(controllerContext.liveOrShuttingDownBrokerIds.toSeq, Set.empty)
    }
```

Reply:
I think we only need to check the return result of `partitionStateMachine.triggerOnlinePartitionStateChange()`. `partitionStateMachine.handleStateChanges(partitionsWithOfflineLeader.toSeq, OfflinePartition)` doesn't generate any  updateMetadataRequest.

Reply:
thanks for the explanation, I think I'm starting to get a better handle on the intended flow.

Let me know if this seems right:

`partitionStateMachine.handleStateChanges(partitionsWithOfflineLeader.toSeq, OfflinePartition)`. This does **not** send LISR/UpdateMetadataRequest.

`partitionStateMachine.triggerOnlinePartitionStateChange()` .This **does** send LISR/UpdateMetadata

`replicaStateMachine.handleStateChanges(newOfflineReplicasNotForDeletion.toSeq, OfflineReplica)` this sends LISR/UpdateMetadata.


So the  `sendUpdateMetadataRequest`  at the end of onReplicasBecomeOffline is meant to cover any cases where  we did not send out an UpdateMetadataRequest.


So if the previous state machine calls did not send any control requests, we should send a control request. I don't think we should care about the value of partitionsWithOfflineLeader because we need to propagate the offline broker to other brokers' metadata anyway.

Reply:
Yes, your understanding is correct.